---
title: "sc_diff_SplicePattern_usage_analysis"
output: html_document
date: "2025-06-23"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(Seurat)
library(pheatmap)

```


# define inputs
```{r}


sample_name = "cordblood_lraa_sc"

sparse_matrix_data_dir = "cordblood_lraa_sc^splice_pattern-sparseM"

umap_cluster_file = "../sc_cluster-guided_LRAA_v0.5.0/cordblood.LRAA.sc_clust_guided.genes-cell_cluster_assignments.wUMAP.tsv.gz"

cluster_pseudobulk_matrix_filename = "cordblood_lraa_sc^splice_pattern-sparseM.cluster_pseudobulk.matrix"

gtf_filename = "cordblood.LRAA_merged.new.gtf.updated.splice_pattern_collapsed.segmented.gtf"

diff_iso_usage_stats_filename = "diff_splice_pattern_usage/diff_splice_pattern_usage.diff_iso.chi2.tsv"

```


# parse inputs

```{r}

isoform_expr_data = Read10X(data.dir=sparse_matrix_data_dir,
               gene.column = 1,
               cell.column = 2,
               unique.features = TRUE,
               strip.suffix = FALSE)

```



```{r}

umap_df = read.csv(umap_cluster_file, header=T, sep="\t")

umap_df %>% head()

```


```{r}

cluster_counts_matrix = read.csv(cluster_pseudobulk_matrix_filename, header=T, row.names=1, sep="\t")


cluster_counts_matrix %>% head()

```

```{r}

cluster_CPM_matrix = sweep(cluster_counts_matrix, 2, colSums(cluster_counts_matrix), "/") * 1e6

head(cluster_CPM_matrix)
```



```{r}
# Read the GTF file
gtf_parsed <- read_tsv( gtf_filename,
                comment = "#", 
                col_names = FALSE, 
                col_types = cols(.default = "c"))

# Assign standard GTF column names
colnames(gtf_parsed)[1:9] <- c("seqname", "source", "feature", "start", "end", 
                        "score", "strand", "frame", "attribute")

# Function to parse attributes into a named list
parse_attributes <- function(attr_str) {
  attrs <- str_split(attr_str, ";\\s*")[[1]]
  kv_pairs <- str_match(attrs, '^(\\S+)\\s+"([^"]+)"')
  kv_pairs <- kv_pairs[!is.na(kv_pairs[, 1]), , drop = FALSE]
  if (nrow(kv_pairs) == 0) return(named(list()))
  set_names(kv_pairs[, 3], kv_pairs[, 2])
}

# Safely parse attributes
gtf_parsed$attr_list <- map(gtf_parsed$attribute, parse_attributes)

# Turn the named list column into separate columns
gtf_parsed <- gtf_parsed %>% unnest_wider(attr_list)

```






# utilty functions

```{r}

################
## UMAP display
################

get_isoform_umap = function(gene_of_interest, restrict_to_transcript_ids = NULL) {
  
  
  if (! is.null(restrict_to_transcript_ids)) {
     transcript_expr_data = data.frame(isoform_expr_data[
       rownames(isoform_expr_data) %in% restrict_to_transcript_ids,])
  } else {
  
    transcript_expr_data = data.frame(isoform_expr_data[grepl(gene_of_interest, rownames(isoform_expr_data)),])
  }
  
   transcript_expr_data$transcript_id = rownames(transcript_expr_data)
   transcript_expr_data = transcript_expr_data %>% gather(key=cell_barcode, value=read_count, -transcript_id) %>%
     filter(read_count >= 1)
  
   umap_df_w_expr_data = right_join(umap_df, transcript_expr_data,
                                   by='cell_barcode')
   
   return(umap_df_w_expr_data)
}


base_umap = umap_df %>% ggplot(aes(x=umap_1, y=umap_2)) + geom_point(color='black', alpha=0.1)

plot_isoform_umap = function(gene_of_interest, restrict_to_transcript_ids = NULL) {
  
  isoform_umap = get_isoform_umap(gene_of_interest, restrict_to_transcript_ids)
  
  base_umap + geom_point(data=isoform_umap, aes(color=transcript_id), alpha=0.5) +
    facet_wrap(~transcript_id) +
    theme_bw() +
    ggtitle(gene_of_interest)  +
     theme(legend.position="none")
  
}


```



```{r}

#####################################
# Gene structure and heatmap display
#####################################

get_gene_structure_matrix = function(gene_of_interest) {
 # Filter to exons only
  exon_df <- gtf_parsed %>%
    filter(grepl(gene_of_interest, gene_id)) %>%
    filter(feature == "exon") %>%
    mutate(
      start = as.integer(start),
      end = as.integer(end),
      exon_coords = paste0(start, "-", end)
    )
  
  # Sort exon coordinates by start position
  exon_levels <- exon_df %>%
    distinct(exon_coords, start) %>%
    arrange(start) %>%
    pull(exon_coords)
  
  # Build binary presence matrix (long format)
  exon_binary_df <- exon_df %>%
    select(transcript_id, exon_coords) %>%
    distinct() %>%
    mutate(present = 1)
  
  # Set factor levels for exon coordinates (x-axis)
  exon_binary_df$exon_coords <- factor(exon_binary_df$exon_coords, levels = exon_levels)
  
  
  # Ensure binary_df is a regular data frame (not grouped or list-columned)
  exon_binary_df <- exon_binary_df %>% ungroup()
  
  # Count number of exons per transcript using tally() and sort by num exons
  transcript_levels <- exon_binary_df %>%
    filter(present == 1) %>%
    group_by(transcript_id) %>%
    tally(name = "n_exons") %>%
    arrange(desc(n_exons)) %>%
    pull(transcript_id)
  
  # Set transcript_id factor levels accordingly
  exon_binary_df$transcript_id <- factor(exon_binary_df$transcript_id, levels = transcript_levels)
  
  return(exon_binary_df)
    
}

gene_structure_heatmap = function(gene_of_interest) {

  exon_binary_df = get_gene_structure_matrix(gene_of_interest)
  
  # Plot heatmap
  p = ggplot(exon_binary_df, aes(x = exon_coords, y = transcript_id, fill = factor(present))) +
    geom_tile(color = "grey80") +
    scale_fill_manual(values = c("1" = "black"), guide = "none") +
    labs(
      title = "Transcript Exon Usage Heatmap",
      x = "Exon Coordinates (start-end)",
      y = "Transcript ID"
    ) +
    theme_minimal(base_size = 10) +
    theme(
      axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
      panel.grid = element_blank()
    )
  
  return(p)

}


get_expression_pheatmap_w_exon_structures = function(gene_of_interest) {

  transcript_ids = gtf_parsed %>% filter(grepl(gene_of_interest, gene_id)) %>% 
    filter(feature == "transcript") %>% select(transcript_id) %>% unique() %>% pull(transcript_id)
  
  isoform_expr = cluster_CPM_matrix[rownames(cluster_CPM_matrix) %in% transcript_ids,]
  
  # Load expression matrix
  expr_mat <- as.matrix(isoform_expr)
  
  # Log transform if desired
  log_expr <- log2(expr_mat + 1)
  
  # Load exon structure info (from earlier script)
  # binary_df has transcript_id, exon_coords, and present (1/0)
  # Ensure transcript IDs match (e.g., same format)
  
  exon_binary_df = get_gene_structure_matrix(gene_of_interest)
  
  # Create exon structure annotation matrix (transcripts x exon_coords)
  exon_mat <- exon_binary_df %>%
    filter(present == 1) %>%
    select(transcript_id, exon_coords) %>%
    mutate(value = "●") %>%
    pivot_wider(names_from = exon_coords, values_from = value, values_fill = "") %>%
    column_to_rownames("transcript_id") %>%
    as.data.frame()
  
  
  
  # Get exon column names and sort by numeric start coordinate
  sorted_exon_cols <- exon_mat %>%
    colnames() %>%
    as_tibble() %>%
    dplyr::rename(coord = value) %>%
    mutate(start = as.integer(str_extract(coord, "^\\d+"))) %>%
    arrange(start) %>%
    pull(coord)
  
  # Reorder the exon matrix columns
  exon_mat_sorted <- exon_mat[, rev(sorted_exon_cols)]
  
  
  # If exon presence is encoded as "●" and "" or 1/0
  # Convert to "yes"/"no" to be explicit
  exon_mat_color <- exon_mat_sorted %>%
    mutate(across(everything(), ~ ifelse(. != "", "yes", "no")))  # or . == 1
  
  # Make all columns factors
  exon_mat_color <- exon_mat_color %>%
    mutate(across(everything(), as.factor))
  
  # Build color mapping for all exon annotation columns
  annotation_colors <- list()
  
  for (col in colnames(exon_mat_color)) {
    annotation_colors[[col]] <- c("yes" = "black", "no" = "white")
  }
  
  # Filter so exon_mat rows match log_expr
  common_ids <- intersect(rownames(log_expr), rownames(exon_mat))
  log_expr <- log_expr[common_ids, ]
  exon_mat_sorted <- exon_mat_sorted[common_ids, ]
  
  
  # Generate heatmap with exon structure annotations as row labels
  pheatmap(log_expr,
           show_rownames = TRUE,
           show_colnames = TRUE,
           cluster_rows = TRUE,
           cluster_cols = TRUE,
           annotation_row = exon_mat_color,
           annotation_colors = annotation_colors,
           fontsize_row = 6,
           main = "Transcript Expression with Exon Structure",
           annotation_legend = FALSE,
  )
  
}

```



```{r}

my_gene_of_interest = "EGFL7"


get_expression_pheatmap_w_exon_structures(my_gene_of_interest)


````



```{r}

get_expression_ggplot2_heatmap_w_exon_structures = function(
    gene_of_interest, 
    min_isoform_frac_expr_any_cluster = 0) {

  library(patchwork)
  library(ggdendro)
  
  
  transcript_ids = gtf_parsed %>% filter(grepl(gene_of_interest, gene_id)) %>% 
    filter(feature == "transcript") %>% select(transcript_id) %>% unique() %>% pull(transcript_id)
  
  isoform_expr = cluster_CPM_matrix[rownames(cluster_CPM_matrix) %in% transcript_ids,]
  
  isoform_frac_expr <- sweep(isoform_expr, 2, colSums(isoform_expr), "/")
  
  if (min_isoform_frac_expr_any_cluster > 0) {
    
    transcript_ids = names(which(rowSums(isoform_frac_expr >= min_isoform_frac_expr_any_cluster) > 0))
     
    if (length(transcript_ids) < 2) {
       stop("Too few isoforms left after filtering based on min isoform fraction") 
    }
    
    isoform_expr = isoform_expr[rownames(isoform_expr) %in% transcript_ids,]
    
    isoform_frac_expr = isoform_frac_expr[rownames(isoform_frac_expr) %in% transcript_ids,]
    
  }
  
  
  # Load expression matrix
  expr_mat <- as.matrix(isoform_expr)
  
  # Log transform if desired
  log_expr <- log2(expr_mat + 1)

  
  col_order <- hclust(dist(t(log_expr)))$order
  ordered_clusters <- colnames(log_expr)[col_order]

  # Convert to long format
  expr_long <- as.data.frame(log_expr) %>%
    rownames_to_column("transcript_id") %>%
    pivot_longer(-transcript_id, names_to = "cluster", values_to = "expression")
  
  frac_expr_long = isoform_frac_expr %>%
    rownames_to_column("transcript_id") %>%
    pivot_longer(-transcript_id, names_to = "cluster", values_to = "expression")
  

  expr_long$cluster <- factor(expr_long$cluster, levels = ordered_clusters)
  frac_expr_long$cluster <- factor(frac_expr_long$cluster, levels = ordered_clusters)

  
  # Cluster rows using base hclust on Euclidean distances
  row_order <- hclust(dist(log_expr))$order
  expr_long$transcript_id <- factor(expr_long$transcript_id, 
                                    levels = rownames(log_expr)[row_order])
  
  frac_expr_long$transcript_id <- factor(frac_expr_long$transcript_id, 
                                    levels = rownames(log_expr)[row_order])
  
  
  exon_binary_df = get_gene_structure_matrix(gene_of_interest)
  
  exon_binary_df = exon_binary_df %>% filter(transcript_id %in% transcript_ids)
  
  # Create exon structure annotation matrix (transcripts x exon_coords)
  exon_mat <- exon_binary_df %>%
    filter(present == 1) %>%
    select(transcript_id, exon_coords) %>%
    mutate(value = "●") %>%
    pivot_wider(names_from = exon_coords, values_from = value, values_fill = "") %>%
    column_to_rownames("transcript_id") %>%
    as.data.frame()
  
  
  
  # Long format for exon annotations
  exon_anno_long <- exon_mat %>%
    rownames_to_column("transcript_id") %>%
    pivot_longer(-transcript_id, names_to = "exon_coords", values_to = "present") %>%
    filter(present != "")  # keep only present exons
  
  # Ensure same transcript order as clustering
  exon_anno_long$transcript_id <- factor(exon_anno_long$transcript_id, 
                                         levels = levels(expr_long$transcript_id))
  
  

  
  # Get dendrogram data
  row_dend <- hclust(dist(log_expr))
  dend <- as.dendrogram(row_dend)
  ddata <- dendro_data(dend, type = "rectangle")
  
  # Get row order for consistency
  row_order <- row_dend$order
  transcript_order <- rownames(log_expr)[row_order]
  
  # Create mapping from x (numeric position) to label
  label_map <- ddata$labels %>%
    mutate(leaf_position = row_number()) %>%
    select(leaf_position, label)
  
  # Create a factor for labels ordered to match heatmap transcript_id
  label_map <- label_map %>%
    mutate(transcript_id = factor(label, levels = transcript_order),
           new_pos = as.numeric(transcript_id))
  
  # Replace x and xend in segments using leaf positions
  segments_fixed <- ddata$segments %>%
    left_join(label_map %>% select(leaf_position, new_pos), by = c("x" = "leaf_position")) %>%
    mutate(x = ifelse(!is.na(new_pos), new_pos, x)) %>%
    select(-new_pos) %>%
    left_join(label_map %>% select(leaf_position, new_pos), by = c("xend" = "leaf_position")) %>%
    mutate(xend = ifelse(!is.na(new_pos), new_pos, xend)) %>%
    select(-new_pos)
  
  p_dend <- ggplot(segments_fixed) +
    geom_segment(aes(x = y, xend = yend, y = x, yend = xend)) +
    scale_x_reverse() +  # ← This flips the tree so root is on the left
    theme_void() +
    theme(plot.margin = margin(t = 10, r = 5, b = 10, l = 10))
  
  
  # Expression heatmap
  p_expr <- ggplot(expr_long, aes(x = cluster, y = transcript_id, fill = expression)) +
    geom_tile() +
    scale_fill_viridis_c() +
    theme_minimal() +
    labs(title = "Iso Expression", x = "Cluster", y = NULL) +
    theme(axis.text.x = element_text(angle = 90, hjust = 1),
          axis.text.y = element_blank(),
          axis.ticks.y = element_blank())
  
  p_frac_expr = ggplot(frac_expr_long, aes(x = cluster, y = transcript_id, fill = expression)) +
    geom_tile() +
    scale_fill_viridis_c() +
    theme_minimal() +
    labs(title = "Iso Fraction", x = "Cluster", y = NULL) +
    theme(axis.text.x = element_text(angle = 90, hjust = 1),
          axis.text.y = element_blank(),
          axis.ticks.y = element_blank())
  
  
  # remove legends
  p_expr <- p_expr + theme(legend.position = "none")
  p_frac_expr <- p_frac_expr + theme(legend.position = "none")

  
  
  
  # Exon structure tile map
  p_exon <- ggplot(exon_anno_long, aes(x = exon_coords, y = transcript_id)) +
    geom_tile(fill = "black") +
    theme_minimal() +
    labs(title = "Exon Structure", x = "Exon (start-end)", y = NULL) +
    theme(axis.text.x = element_text(angle = 90, hjust = 1),
          axis.text.y = element_text(size = 6),
          plot.margin = margin(t = 10, r = 10, b = 10, l = 0))
  
  
  
  g_dend <- ggplotGrob(p_dend)
  g_exon <- ggplotGrob(p_exon)
  g_expr <- ggplotGrob(p_expr)
  g_frac_expr <- ggplotGrob(p_frac_expr)
  
  
  common_heights <- grid::unit.pmax(g_dend$heights,
                                    g_exon$heights,
                                    g_expr$heights,
                                    g_frac_expr$heights)
  
  g_dend$heights <- common_heights
  g_exon$heights <- common_heights
  g_expr$heights <- common_heights
  g_frac_expr$heights <- common_heights
  

  
  combined_plot <- 
    wrap_elements(g_dend) +
    wrap_elements(g_exon) +
    wrap_elements(g_expr) +
    wrap_elements(g_frac_expr) +
    plot_layout(ncol = 4, widths = c(0.5, 1.2, 0.5, 0.5))
  
  return(list(
    plot = combined_plot, 
    transcript_ids = transcript_ids)
    )
  
}

```


```{r}

gene_exon_heatmap_plot = get_expression_ggplot2_heatmap_w_exon_structures(my_gene_of_interest, 
                                                                0.1)

p_exon_expr = gene_exon_heatmap_plot$plot

p_exon_expr
```


```{r}

p_umap = plot_isoform_umap(my_gene_of_interest, gene_exon_heatmap_plot$transcript_ids)

p_umap

```

```{r}

library(cowplot)

p_both = plot_grid(p_exon_expr, p_umap, ncol=1)

ggsave(p_both, 
       file=paste0(my_gene_of_interest, ".pdf"),
       width=12, 
       height=12
       )


```

```{r}

library(cowplot)

make_diff_iso_usage_compound_plot = function(gene_of_interest, min_iso_fraction = 0) {
  
  p_exon_expr_info = get_expression_ggplot2_heatmap_w_exon_structures(gene_of_interest, min_iso_fraction)
  
  p_exon_expr = p_exon_expr_info$plot
  
  p_umap = plot_isoform_umap(gene_of_interest, p_exon_expr_info$transcript_ids)
  
  p_both = plot_grid(p_exon_expr, p_umap, ncol=1)
  
  return(p_both)
  
}

p_my_gene_of_interest = make_diff_iso_usage_compound_plot(my_gene_of_interest, 0.1)

ggsave(p_my_gene_of_interest, 
       file=paste0(my_gene_of_interest, ".pdf"),
       width=12, 
       height=12
       )

p_my_gene_of_interest

```

#############################
# Make plots for top examples


```{r}

# Specify the path for the directory you want to create
num_top_diff_iso_usage_plots = 100
top_diff_iso_usage_plots_dir <- paste0("top_", num_top_diff_iso_usage_plots, "_diff_iso_usage_plots_minIsoF0.1")

# Check if the directory already exists
if (!dir.exists(top_diff_iso_usage_plots_dir)) {
  # If the directory doesn't exist, create it
  dir.create(top_diff_iso_usage_plots_dir)
}


```

```{r}

diff_iso_usage_stats = read.csv(diff_iso_usage_stats_filename, sep="\t", header=T)

diff_iso_usage_stats %>% head()


```

```{r}
# get top candidates.

top_diff_iso_usage_genes = diff_iso_usage_stats %>% filter(significant == "True") %>% 
  group_by(gene_id) %>% 
      # get top entruy per gene
      arrange(adj_pvalue, desc(abs(delta_pi))) %>% filter(row_number() == 1) %>% ungroup() %>%
  arrange(adj_pvalue, desc(abs(delta_pi))) %>%
  head(num_top_diff_iso_usage_plots) %>%
  pull(gene_id) 

top_diff_iso_usage_genes

```

```{r}


counter = 0
for(gene_sym_value in top_diff_iso_usage_genes) {
  
    message("plotting ", counter, " gene: ", gene_sym_value)
  
    filename = paste0(top_diff_iso_usage_plots_dir, "/", counter, ".", gene_sym_value, ".diff_iso_usage.pdf")
    
    counter = counter + 1
    
    print(counter)
    
    #if (counter < 200) {
    #   next;
    #}
    
    gene_sym_value_regex = paste0("^", gene_sym_value, "\\^")
    
    tryCatch({
      
        pdf(filename)
    
        p = make_diff_iso_usage_compound_plot(gene_sym_value_regex, 0.1)
        plot(p)
        dev.off()
      
     }, error = function(e) {
      message("Caught error: ", e$message)
  
    })
}  




```





